### PR TITLE
[WIP, DNM] crimson: poseidonstore initial commit

### PR DESCRIPTION
**Pull request by myoungwon**
Based on the document (https://docs.ceph.com/en/latest/dev/crimson/poseidonstore/), 
the purpose of this PR is to implement WAL and basic cache.

Signed-off-by: Myoungwon Oh <myoungwon.oh@samsumg.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by sebastian-philipp** [comment]
jenkins test make check

**Comment by myoungwon** [comment]
@athanatos Would you mind looking at this? It seems that we need your comment before we make further updates on this.

File: src/crimson/os/poseidonstore/poseidonstore.h [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+ * partion 1 means a sharded partition
+ * 
+ * ** Discussion
+ * The best way to achieve high IOPS is that each sharded service (reactor thread) manages only 
+ * a sharded partition. With this policy; each sharded partion handles pre-configured PGs,
+ * if the number of PGs grows, the number of sharded parition can also increase. However, what happens if
+ * the number of sharded services (reactor thread) grows or shrinks with the same PGs? 
+ * Do we need to consider this case?
+ * To be conservative, the following design assumes that we need a lock to protect two reactor thread
+ * access a shared partition at a time, considering such case.
```
Comment (athanatos) [inline src/crimson/os/poseidonstore/poseidonstore.h:59]: This discussion really isn't useful here.  Please replace it with a concise, exact description of what you are doing.

**Comment by athanatos** [comment]
What specifically would you like feedback on?  It looks like there's a lot that isn't implemented yet.  This, of course, is fine but it means that I'm not sure what  you want me to look at.  One note I have is that the gross structure seems pretty similar to seastore.  Are you certain it wouldn't be more efficient/easier to modify seastore?

Also, remind me, was the goal here purely persistent memory?

**Comment by myoungwon** [comment]
> What specifically would you like feedback on?

We are at the stage of low-level design (LLD) of each component in PoseidonStore.
This PR includes the LLD of WAL and cache components (whose implementations have not yet been completed).
**_We would like to kindly ask a feedback from you on the LLD of the two components_**. 
Do you think they make sense? Are there any missing points to be discussed before we continue implementing?

Also, **_we were wondering what the initial PR should include._**
Should it include 1) the LLD of EVERY component (rather than a set of components) and 2) all the definitions of main data strucutures and functions?
Or is it ok if we start from one component in a component-by-component manner just like we did?

> One note I have is that the gross structure seems pretty similar to seastore. Are you certain it wouldn't be more efficient/easier to modify seastore?

For WAL and cache components, we have not found any reasons that we start from scratch, especially where SeaStore has well-defined data structures built for the same purposes
as PoseidonStore's. 
However, those components of PoseidonStore will certainly do have a number of functions that have to be implemented differently due to the different design decisions from SeaStore. (i.e., hybrid update strategies (in-place + out-of-place))
Although the current PR does not cover them, we will eventually add them in next commits.
For this reason, we chose to borrow the SeaStore's data structures, rather than to create a new branch from SeaStore.


> Also, remind me, was the goal here purely persistent memory?

No, our target is high-end NVMe SSD (e.g., Z-SSD). Note that in the current implementation, we use memory as a block device for test purpose.

**Comment by athanatos** [comment]
Well, we do plan on enabling seastore to do in-place updates for optane ssds.  The architectural similarity rather suggests that adapting seastore wouldn't be particularly difficult and would have the benefit of eventually enabling blended deployments using zoned devices with faster optane/Z-SSD as faster tiers.  As I said when you initially described this, I'm still sceptical that introducing another implementation is wise.

I need to find time to look at this PR in detail, I'll try to get to it this week.

**Comment by myoungwon** [comment]
The tiering approach in object store sounds great and we agree that a single implementation has several advantages over multiple independent implementations.
However, our design also includes some other decisions that may not be compatible with the current Seastore's implementation.; for example,  sharded on-disk layout, use of io_uring, and consideration of atomic write unit.
Thus, at this stage, we thought that it would be better to develop SeaStore and PoseidonStore separately in parallel but with the future integration in mind. Regarding this, we need to figure out what components can be shared, what can be implemented on top of SeaStore's components, and what must be implemented from scratch.

**Comment by athanatos** [comment]
The atomic write unit portion is simply a journaling optimization, right?  I think it be really easy to make the journal portion of seastore an interface with this as an additional implementation for cases where the underlying hardware supports it.

Seastore plans to have a per-reactor partitioned layout with any particular pg localized to a single partition.

I'm not sure how io_uring enters into it, what's the plan there?

From a strategy standpoint, I don't think it makes any sense at all to start separately if the end goal is to unify them.  I think everything I've seen so far would fit within the TransactionManager layer of seastore (though I haven't had time to review this PR in detail yet).  Doing that would allows you to immediately leverage the onode index and other higher level structures being worked on for seastore -- that seems like an obviously huge win.

File: src/crimson/os/poseidonstore/cache.h [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+ *
+ * As initail step, we borrow seastore's data structure to read/write 
+ * the data to the cache.
+ * Similar to seastore, Cache will maintain general info for entire cache.
+ * CacheExtent is a cache entry and any cache contents such as TestBlock, 
+ * Onode and Block can be created inherited from CacheExtent.
+ *
+ *
+  *
+ */
```
Comment (athanatos) [inline src/crimson/os/poseidonstore/cache.h:91]: I don't see any reason why these design choices couldn't be implemented right now in seastore.  Regarding caching, the plan for seastore is to have per-reactor partitions with their own allocation/metadata structures with each PG localized to exactly one (transactions won't span multiple PGs).

Comment (athanatos) [inline src/crimson/os/poseidonstore/cache.h:91]: I don't see any reason why these design choices couldn't be implemented right now in seastore. Regarding sharding, the plan for seastore is to have per-reactor partitions with their own allocation/metadata structures with each PG localized to exactly one (transactions won't span multiple PGs).

File: src/crimson/os/poseidonstore/cache.cc [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+      bl.append(c->get_bptr(), c->modified_range.range_start(), 
+		    c->modified_range.range_end());
+      record->to_write.push_back(
+	  write_item{
+	    std::move(bl),
+	    c->get_type(),
+	    c->get_laddr()
+	  });
+    }
+  }
```
Comment (athanatos) [inline src/crimson/os/poseidonstore/cache.cc:65]: This is a departure from seastore -- currently we assume that "deltas" will be replayed in full from the initial state of the extent.  This variant, however, is something we plan on introducing in the not particularly far future to permit in-place modification for media where that is more expedient.

File: src/crimson/os/poseidonstore/wal.h [hunk 4]
```diff
@@ -0,0 +1,453 @@
...
+   *
+   * Will write bl from offset to length
+   * 
+   * @param bufferlist to write
+   * @param laddr_t where data is written
+   * @return <relative address to write,
+   * 	      total length written to WAL (bytes)>
+   *
+   */
+  write_block_ret append_block(ceph::bufferlist bl, laddr_t addr);
```
Comment (athanatos) [inline src/crimson/os/poseidonstore/wal.h:85]: So, what does laddr_t mean for poseidon store?  poseidonstore_types.h mentions that it's the logical address from the pov of the onode, but that doesn't really make sense here.  Why would the wal know about that?  Which onode?

**Comment by athanatos** [comment]
So, I've looked over this a bit.  The way you want to do caching (the basic approach described in the comment) seems great and super easy to port to seastore.  I'd actively like to add the update pattern here (write ahead modified regions to journal and then directly update on-disk regions) to seastore to make use of devices where direct updates are cheap.  I wasn't yet able to fully grasp some details of the wal, but I'm entirely fine with adding an alternate journal implementation making use of atomic write and other hardware features where supported.

In short, I just don't yet see anything here preventing us from simply introducing these modifications to seastore.  Absent some compelling reason why we can't, I think we'll move far far more quickly by introducing these changes into seastore than if we create a second implementation.

**Comment by pjbear1215** [comment]
> The atomic write unit portion is simply a journaling optimization, right? I think it be really easy to make the journal portion of seastore an interface with this as an additional implementation for cases where the underlying hardware supports it.

Yes, it is basically an optimization that enables skipping issuing flush command or the journaling itself where the underlying NVMe SSD supports a large atomic write unit.

> I'm not sure how io_uring enters into it, what's the plan there?

We plan on using io_uring instead of SPDK. We are not sure that SPDK is the right solution for CPU-efficient I/O processing, especially in a Crimson's threading model, where each reactor thread is responsible for everything for sharded data; message exchange, issuing and completing I/O requests, and etc. SPDK might be CPU efficient only with a specific thread model where one I/O thread is dedicated to issuing and completing I/Os for multiple drives on behalf of other worker threads; it is obviously different from the Crimson's.

In our evaluation (4KB random I/O with NVMe SSDs), io_uring with the default interrupt mode provides the best IOPS per core in a Crimson-like threading model (among SPDK, AIO, and io_uring); SPDK can provide up to 20% higher IOPS but at the cost of up to 4x higher CPU consumption, and the performance benefit becomes negligible as the I/O performance is saturated. Other than CPU-efficient I/O processing, io_uring also provides SQE chaining feature that enables efficient transaction processing. For example, we don't have to wait for flushing before writing commit block; we can chain a series of writes that must be executed in order. Due to these advantages of io_uring, we've considered io_uring as the default I/O interface for PoseidonStore.

**Comment by myoungwon** [comment]
We agree that introducing our ideas into SeaStore makes implementation much easier by leveraging the SeaStore's data structures, and we are fine with that; we definitely want to make it widely used in industry as soon as possible.
But let's clear up this. If we understand correctly, there will be three modes in a future SeaStore release as follows: 
1) Original SeaStore for ZNS 
2) PoseidonStore for high-end SSDs 
3) Tiered Approach - PoseidonStore (fast tier) + Original Seastore (slower tier). 
Is this correct? We hope that PoseidonStore also can work alone although it is implemented within SeaStore.

**Comment by athanatos** [comment]
I expect that the internals of seastore will, depending on hardware and configuration, select appropriate internal implementations.  From what I see here, that would mean a high speed ssd specific journal implementation and a transaction/caching/allocation mode where we do direct write back after journal writeahead (though I want us to think through ensuring that the architecture will naturally handle tiered/combination configurations as well where appropriate).   In situations where there is only a high speed ssd, I'd expect the IO pathway to work basically as you describe, but I want the components to be integrated as cleanly as possible into seastore both to permit tiered/combination configurations but mostly to reduce ongoing maintenance and speed implementation.  I'm happy to discuss any situiations where we need to generalize/adjust choices in seastore to fit better, but I really don't want to see similar, parallel implementations.

I'm with you on io_uring -- my kernel side colleagues appear to prefer it and have serious reservations about SPDK.  There's a patch series around the seastar upstream to integrate it into seastar, at which point we can integrate it into seastore.

**Comment by myoungwon** [comment]
We're not clear what to do at this stage.
It seems that having a discussion
is required first to clarify which items need to be implemented and generalized.

I'm just listing up the discussion items in my mind as belows.

1. Applying WAL in seastore
Basically, the Journal in seastore appends segments in order, considering delta.
However, WAL in poseidonstore is located at a fixed location, and it is a kind of circular queue.
Also, it needs to take into account committed and applied state, considering moving the wal entry 
to the data partition.
We thought some basic data structures can be reused, but the Journal in seastore need to be generalized
to cover the WAL's characteristics. 
Which interfaces are needed to cover both stores?

Journal class 
- submit_record
- init
- replay (from ~ to)
- encode/decode
...

2. do_transaction
The works during do_transaciotn() differs between seastore and poseidonstore;
seastore might call gc unlike poseidonstore's delay work for applying data 
even though both have common works (e.g., reading metadata and submit_entry).
How can we generalize do_transaction()?

3. Data partition
Unlike seastore, Poseidonstore has both WAL and the data partition, which means this item probably requires 
a separated class.

4. Block cache
This can be reusable. We probably can apply our idea to seastore by inhering from the existing cache design.
